query/src/main/scala/filodb/query/exec/rangefn/RangeFunction.scala

```diff
@@ -225,82 +226,128 @@ trait ChunkedLongRangeFunction extends TimeRangeFunction[TransientRow] {
-    case Some(StdDevOverTime) => () => new StdDevOverTimeChunkedFunctionL
-    case Some(StdVarOverTime) => () => new StdVarOverTimeChunkedFunctionL
-    case _                    => iteratingFunction(func, funcParams)
+  def longChunkedFunction(dataset: Dataset,
+                          func: Option[RangeFunctionId],
+                          funcParams: Seq[Any] = Nil): RangeFunctionGenerator = {
+    func match {
+      case None                 => () => new LastSampleChunkedFunctionL
+      case Some(CountOverTime)  => () => if (dataset.hasDownsampledData) new SumOverTimeChunkedFunctionL
+                                         else new CountOverTimeChunkedFunction()
+      case Some(SumOverTime)    => () => new SumOverTimeChunkedFunctionL
+      case Some(AvgOverTime)    => () => if (dataset.hasDownsampledData) {
+                                           val cntCol = dataset.dataColId("count")
+                                           new AvgWithSumAndCountOverTimeFuncL(cntCol)
+                                         }
+                                         else new AvgOverTimeChunkedFunctionL
+      case Some(MinOverTime)    => () => new MinOverTimeChunkedFunctionL
+      case Some(MaxOverTime)    => () => new MaxOverTimeChunkedFunctionL
+      case Some(StdDevOverTime) => () => new StdDevOverTimeChunkedFunctionL
+      case Some(StdVarOverTime) => () => new StdVarOverTimeChunkedFunctionL
+      case _                    => iteratingFunction(func, funcParams)
+    }
   }
 
   /**
    * Returns a function to generate a ChunkedRangeFunction for Double columns
    */
-  def doubleChunkedFunction(func: Option[RangeFunctionId],
+  def doubleChunkedFunction(dataset: Dataset,
+                            func: Option[RangeFunctionId],
                             config: QueryConfig,
-                            funcParams: Seq[Any] = Nil): RangeFunctionGenerator = func match {
-    case None                 => () => new LastSampleChunkedFunctionD
-    case Some(Rate)     if config.has("faster-rate") => () => new ChunkedRateFunction
-    case Some(Increase) if config.has("faster-rate") => () => new ChunkedIncreaseFunction
-    case Some(Delta)    if config.has("faster-rate") => () => new ChunkedDeltaFunction
-    case Some(CountOverTime)  => () => new CountOverTimeChunkedFunctionD()
-    case Some(SumOverTime)    => () => new SumOverTimeChunkedFunctionD
-    case Some(AvgOverTime)    => () => new AvgOverTimeChunkedFunctionD
-    case Some(MinOverTime)    => () => new MinOverTimeChunkedFunctionD
-    case Some(MaxOverTime)    => () => new MaxOverTimeChunkedFunctionD
-    case Some(StdDevOverTime) => () => new StdDevOverTimeChunkedFunctionD
-    case Some(StdVarOverTime) => () => new StdVarOverTimeChunkedFunctionD
-    case _                    => iteratingFunction(func, funcParams)
+                            funcParams: Seq[Any] = Nil): RangeFunctionGenerator = {
```

Comment:
Again keep the old indentation the way it is, why change it?

Reply:
It makes the changes much harder to see

Reply:
sure. Intellij.

Reply:
I also had to get past the line width limitation.